sqlstats: generalize insight.ConcurrentBufferIngester

xinhaoz opened the pull request:
This commit copies the ConcurrentBufferIngester from insights into the sqlstats
pkg in preparation to have it consume the sql execution event and handle the
writing to both sqlstats and insights systems, instead of just for insights. At
this stage we are preserving how this ingester consumes events. The ingester is
still only used for insights writing.

Previously, the insights registry had the responsibility of buffering statements
by session id and processing them once the entire transaction completes (via
receicing a transaction event). The responsibility of buffering statements by
session has now moved to the ingester. As a result we remove the
`ObserveStatement` function from the insights registry which was only used to
insert into the statement buffer.

Part of: https://github.com/cockroachdb/cockroach/issues/141024

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143768)
<!-- Reviewable:end -->

xinhaoz committed:
sqlstats: generalize insight.ConcurrentBufferIngester

This commit copies the ConcurrentBufferIngester from insights into the sqlstats
pkg in preparation to have it consume the sql execution event and handle the
writing to both sqlstats and insights systems, instead of just for insights. At
this stage we are preserving how this ingester consumes events. The ingester is
still only used for insights writing.

Previously, the insights registry had the responsibility of buffering statements
by session id and processing them once the entire transaction completes (via
receicing a transaction event). The responsibility of buffering statements by
session has now moved to the ingester. As a result we remove the
`ObserveStatement` function from the insights registry which was only used to
insert into the statement buffer.

Part of: #141024

Epic: none
Release note: None

xinhaoz commented on pkg/sql/sqlstats/sslocal/sql_stats_ingester_test.go:
Tests previously from `ingester_test.go`.

xinhaoz committed:
sqlstats,insights: adjust testing and move insights knobs

Now that the sql exec stats ingester lives in the sqlstats package we move the
relevant testing knobs from the insights package to the sql stats knobs. This
commit also adds the `OnIngesterFlush` testing knob to synchronously test the
ingester flush.

Epic: none
Release note: None

kyle-a-wong reviewed (approved):
lgtm!

xinhaoz commented:
TFTR!
bors r+
